solidity/contracts/libs/RateLimited.sol

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity >=0.8.0;
+import {OwnableUpgradeable} from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+
+/**
+ * @title RateLimited
+ * @notice A contract used to keep track of an address sender's token amount limits.
+ * @dev Implements a modified token bucket algorithm where the bucket is full in the beginning and gradually refills
+ * See: https://dev.to/satrobit/rate-limiting-using-the-token-bucket-algorithm-3cjh
+ **/
+contract RateLimited is OwnableUpgradeable {
+    uint256 public constant DURATION = 1 days; // 86400
+    uint256 public filledLevel; /// @notice Current filled level
+    uint256 public refillRate; /// @notice Tokens per second refill rate
+    uint256 public lastUpdated; /// @notice Timestamp of the last time an action has been taken TODO prob can be uint40
+
+    event RateLimitSet(uint256 _oldCapacity, uint256 _newCapacity);
+
+    constructor(uint256 _capacity) {
+        _transferOwnership(msg.sender);
+        setRefillRate(_capacity);
+        filledLevel = _capacity;
+    }
+
+    error RateLimitExceeded(uint256 newLimit, uint256 targetLimit);
+
+    /**
+     * @return The max capacity where the bucket will no longer refill
+     */
+    function maxCapacity() public view returns (uint256) {
+        return refillRate * DURATION;
+    }
+
+    /**
+     * Calculates the refilled amount based on time and refill rate
+     *
+     * Consider an example where there is a 1e18 max token limit per day (86400s)
+     * If half of the tokens has been used, and half a day (43200s) has passed,
+     * then there should be a refill of 0.5e18
+     *
+     * To calculate:
+     *   Refilled = Elapsed * RefilledRate
+     *   Elapsed = timestamp - Limit.lastUpdated
+     *   RefilledRate = Capacity / DURATION
+     *
+     *   If half of the day (43200) has passed, then
+     *   (86400 - 43200) * (1e18 / 86400)  = 0.5e18
+     */
+    function calculateRefilledAmount() internal view returns (uint256) {
+        uint256 elapsed = block.timestamp - lastUpdated;
+        return elapsed * refillRate;
+    }
+
+    /**
+     * Calculates the adjusted fill level based on time
+     */
+    function calculateFilledLevel() public view returns (uint256) {
+        uint256 _capacity = maxCapacity();
+        require(_capacity > 0, "RateLimitNotSet");
+
+        if (lastUpdated + DURATION > block.timestamp) {
```

Comment:
its much easier for my brain to read, idk if its just me
```suggestion
        if (block.timestamp < lastUpdated + DURATION) {
```

Reply:
also I might restructure this to be
```
if *not* this case {
  return max capacity
}
// within window stuff
```

Reply:
Fixed